source/switch.cpp

```diff
@@ -0,0 +1,83 @@
+#include "switch.hpp"
+
+#include <spdlog/spdlog.h>
+
+#include <memory>
+#include <stdexcept>
+
+#include "link.hpp"
+
+#define LOG_WARN(msg) spdlog::warn("{} ({}:{})", msg, __FILE__, __LINE__)
+
+namespace sim {
+Switch::Switch() : m_router(std::make_unique<RoutingModule>()) {}
+
+void Switch::add_inlink(std::shared_ptr<ILink> link) {
+    if (link == nullptr) {
+        LOG_WARN("Add nullptr inlink to switch device");
+        return;
+    }
+    if (link->get_to().get() != this) {
+        LOG_WARN("Inlink destination is not our device");
+        return;
+    }
+    m_router->add_inlink(link);
+}
+
+void Switch::update_routing_table(std::shared_ptr<IRoutingDevice> dest,
+                                  std::shared_ptr<ILink> link) {
+    if (dest == nullptr) {
+        LOG_WARN("Destination device does not exists");
+        return;
+    }
+    if (link == nullptr) {
+        LOG_WARN("Link does not exists");
+        return;
+    }
+    if (link->get_from().get() != this) {
+        LOG_WARN("Link source is not our device");
+        return;
+    }
+    m_router->update_routing_table(dest, link);
+}
+
+std::shared_ptr<ILink> Switch::next_inlink() { return m_router->next_inlink(); }
+
+std::shared_ptr<ILink> Switch::get_link_to_destination(
+    std::shared_ptr<IRoutingDevice> dest) const {
+    return m_router->get_link_to_destination(dest);
+}
+
+DeviceType Switch::get_type() const { return DeviceType::SWITCH; }
+
+void Switch::process() {
+    std::shared_ptr<ILink> link = next_inlink();
+
+    if (link == nullptr) {
+        LOG_WARN("No next inlink");
+        return;
+    }
+
+    std::optional<Packet> optional_packet = link->get_packet();
+    if (!optional_packet.has_value()) {
+        LOG_WARN("No packet in link");
+        return;
+    }
+    Packet packet = optional_packet.value();
+    if (packet.flow == nullptr) {
+        LOG_WARN("No flow in packet");
+        return;
+    }
+    std::shared_ptr<IReceiver> destination =
+        std::shared_ptr<IReceiver>(packet.flow->get_destination());
```

Comment:
Is the call of shared_ptr constructor necessary here?
```suggestion
    std::shared_ptr<IReceiver> destination = packet.flow->get_destination();
```

Reply:
Thanks, it is not secessary